File: Jal/Al/src/C4/U3/KruskalMST.java
package C4.U3;

import C1.Unit5_UnionFind.UF;
import C2.U4.MinPQ;
import C1.Unit3_Stacks.Queue;
import C4.U3.Edge;
import C4.U3.EdgeWeightedGraph;


public class KruskalMST {
    private Queue<Edge> mst;
    private double weight;

    public KruskalMST(EdgeWeightedGraph G)
    {
        mst = new Queue<Edge>();
        MinPQ<Edge> pq = new MinPQ<Edge>(G.edges());
        UF uf = new UF(G.V());

        while(!pq.isEmpty() && mst.size() < G.V() - 1)
        {
            // Get min weight edge on pq and its vertices.
            Edge e = pq.delMin();
            int v = e.either(), w = e.other(v);
            // Ignore ineligible edges.
            if(uf.connected(v, w))
                continue;
            // Merge components.
            uf.union(v, w);
            // Add edge to mst.
            mst.enqueue(e);
            weight += e.weight();
        }
    }

    public Iterable<Edge> edges()
    {
        return mst;
    }

    public double weight()
    {
        return weight;
    }

    public static void main(String[] args)
    {
        StdLib.In in = new StdLib.In(args[0]);
        EdgeWeightedGraph G;
        G = new EdgeWeightedGraph(in);

        KruskalMST mst = new KruskalMST(G);
        for(Edge e : mst.edges())
            StdLib.StdOut.println(e);
        StdLib.StdOut.println(mst.weight());
    }

}

File: Cal/Al/src/C4/U3/KruskalMST.cpp
#include "C4/U3/KruskalMST.h"
#include "C2/U4/MinPQ.h"
#include "C1/Unit5_UnionFind/UF.h"

#include <vector>
#include <iterator>

KruskalMST::KruskalMST(EdgeWeightedGraph* G)
{
    weight = 0;
    vector<Edge*> myEdges;
    G->getEdges(myEdges);
    MinPQ* pq = new MinPQ(G->getE());
    for(vector<Edge*>::iterator it = myEdges.begin(); it != myEdges.end(); it++)
    {
        pq->insert(*it);
    }
    UF* uf = new UF(G->getV());

    while(!pq->isEmpty() && mst.size() < G->getV() - 1)
    {
        // Get min weight edge on pq and its vertices.
        Edge* e = dynamic_cast<Edge*>(pq->delMin());
        int v = e->either(), w = e->other(v);
        // Ignore ineligible edges.
        if(uf->connected(v, w))
            continue;
        // Merge components.
        uf->unionF(v, w);
        // Add edge to mst.
        mst.enqueue(e);
        weight += e->getWeight();
    }

}

void KruskalMST::getEdges(Queue<Edge*>& qEdges)
{
    while(!mst.isEmpty())
    {
        qEdges.enqueue(mst.dequeue());
    }
}

double KruskalMST::getWeight()
{
    return weight;
}
